Sources/TestsCommon/CommonTestUtils.swift

```diff
@@ -122,6 +122,73 @@ open class MongoSwiftTestCase: XCTestCase {
     }
 }
 
+/// Enumerates the different topology configurations that are used throughout the tests
+public enum TestTopologyConfiguration: String, Decodable {
+    case sharded
+    case replicaSet = "replicaset"
+    case single
+
+    /// Determines the topologyType of a client based on the reply returned by running an isMaster command
+    public init(isMasterReply: BSONDocument) throws {
+        // Check for symptoms of different topologies
+        if isMasterReply["msg"] != "isdbgrid" &&
+            isMasterReply["setName"] == nil &&
+            isMasterReply["isreplicaset"] != true {
+            self = .single
+        } else if isMasterReply["msg"] == "isdbgrid" {
+            self = .sharded
+        } else if isMasterReply["ismaster"] == true && isMasterReply["setName"] != nil {
+            self = .replicaSet
+        } else {
+            fatalError("Invalid test topology configuration")
+        }
+    }
+}
+
+/// Enumerates different possible unmet requirements that can be returned by meetsRequirements
+public enum UnmetRequirements {
+    case minServerVersion(actual: ServerVersion, required: ServerVersion)
+    case maxServerVersion(actual: ServerVersion, required: ServerVersion)
+    case topology(actual: TestTopologyConfiguration, required: [TestTopologyConfiguration])
+}
+
+/// Struct representing conditions that a deployment must meet in order for a test file to be run.
+public struct TestRequirement: Decodable {
+    private let minServerVersion: ServerVersion?
+    private let maxServerVersion: ServerVersion?
+    private let topology: [TestTopologyConfiguration]?
+
+    public init(
+        minServerVersion: ServerVersion? = nil,
+        maxServerVersion: ServerVersion? = nil,
+        acceptableTopologies: [TestTopologyConfiguration]? = nil
+    ) {
+        self.minServerVersion = minServerVersion
+        self.maxServerVersion = maxServerVersion
+        self.topology = acceptableTopologies
+    }
+
+    /// Determines if the given deployment meets this requirement.
+    public func isMet(by version: ServerVersion, _ topology: TestTopologyConfiguration) -> UnmetRequirements? {
```

Comment:
something that is really emphasized in Swift API design is clarity at the point of use of a method. this method was originally named `isMet(by:)` which I think made sense given that it returned a boolean. however, I wonder if we can rename it now to make it clearer that what we are returning is an unmet requirement, if one exists?

the same applies to the specific `meetsRequirements` helpers we have for the sync and async modules; when I see e.g. 
```swift
 let unmetRequirements = try client.checkRequirements(testRequirements)
            guard unmetRequirements == nil else {
```
in a test it's not immediately apparent that a nil return value from `checkRequirements` corresponds to "all requirements are met"; you just have to know that, or go look at the method to see.

a first stab at a new name would be `getUnmetRequirement` or something along those lines - I think in that case it's more clear that we either return an unmet requirement, or that we return nil there is no unmet requirement.

relevant reading if you haven't seen this yet: https://swift.org/documentation/api-design-guidelines/

Reply:
oh okay yeah this makes a lot of sense. Thanks for linking them the api design guidelines. I renamed both `checkRequirements` and `isMet` to `getUnmetRequirement`, is it confusing/bad that they have the same name?

Reply:
great, thanks! I think the overlapping names are fine.